tilavarauspalvelu/tasks.py

```diff
@@ -711,6 +712,74 @@ def anonymize_old_users_task() -> None:
     retry_backoff=True,
 )
 def delete_pindora_reservation(reservation_uuid: str) -> None:
+    """
+    Task that can be used to retry a Pindora reservation deletion if it fails
+    in the endpoint. This should only be called if the access code is know to
+    be inactive, since this task may also fail to delete the reservation if
+    Pindora is down for an extended period of time.
+    """
     from tilavarauspalvelu.integrations.keyless_entry import PindoraClient
 
     PindoraClient.delete_reservation(reservation=uuid.UUID(reservation_uuid))
+
+
+@app.task(name="create_missing_pindora_reservations")
+def create_missing_pindora_reservations() -> None:
+    """If a reservation failed to be created in Pindora, e.g. in the staff create endpoint, retry it here."""
+    from tilavarauspalvelu.integrations.keyless_entry import PindoraClient
+
+    reservations: list[Reservation] = list(
+        Reservation.objects.filter(
+            state=ReservationStateChoice.CONFIRMED,
+            access_type=AccessType.ACCESS_CODE,
+            access_code_generated_at=None,
+            end__gt=local_datetime(),
+        )
+    )
+
+    if not reservations:
+        return
+
+    for reservation in reservations:
+        is_active = reservation.access_code_should_be_active
+
+        with suppress(Exception):
+            response = PindoraClient.create_reservation(reservation=reservation, is_active=is_active)
+            reservation.access_code_generated_at = response["access_code_generated_at"]
+            reservation.access_code_is_active = response["access_code_is_active"]
+
+    Reservation.objects.bulk_update(reservations, ["access_code_generated_at", "access_code_is_active"])
+
+
+@app.task(name="update_reservation_is_activate")
+def update_reservation_is_activate() -> None:
+    """If a reservation's access code active state failed to update in Pindora, retry the change here."""
+    from tilavarauspalvelu.integrations.keyless_entry import PindoraClient
+    from tilavarauspalvelu.integrations.keyless_entry.exceptions import PindoraNotFoundError
+
+    reservations: list[Reservation] = list(
+        Reservation.objects.filter(
```

Comment:
Also limit to `access_type=AccessType.ACCESS_CODE`

Reply:
On second thought, I suppose that is not necessary or even desirable, because if somehow access_type is something else but the reservation still has `access_code_generated_at`, we wan't to ensure that it is removed from Pindora.

Reply:
Was going to write that, agreed.